src/workerd/util/sqlite.c++

```diff
@@ -0,0 +1,835 @@
+}
+
+Sqlite::Query::ValuePtr Sqlite::Query::getValue(uint column) {
+  switch (sqlite3_column_type(statement, column)) {
+    case SQLITE_INTEGER: return getInt64(column);
+    case SQLITE_FLOAT:   return getDouble(column);
+    case SQLITE_TEXT:    return getText(column);
+    case SQLITE_BLOB:    return getBlob(column);
+    case SQLITE_NULL:    return nullptr;
+  }
+  KJ_UNREACHABLE;
+}
+
+kj::ArrayPtr<const byte> Sqlite::Query::getBlob(uint column) {
+  const byte* ptr = reinterpret_cast<const byte*>(sqlite3_column_blob(statement, column));
+  return kj::arrayPtr(ptr, sqlite3_column_bytes(statement, column));
+}
+
+kj::StringPtr Sqlite::Query::getText(uint column) {
+  const char* ptr = reinterpret_cast<const char*>(sqlite3_column_text(statement, column));
+  return kj::StringPtr(ptr, sqlite3_column_bytes(statement, column));
+}
+
+int Sqlite::Query::getInt(uint column) {
+  return sqlite3_column_int(statement, column);
+}
+
+int64_t Sqlite::Query::getInt64(uint column) {
+  return sqlite3_column_int64(statement, column);
+}
+
+double Sqlite::Query::getDouble(uint column) {
+  return sqlite3_column_double(statement, column);
+}
+
+bool Sqlite::Query::isNull(uint column) {
+  return sqlite3_column_type(statement, column) == SQLITE_NULL;
+}
+
+// =======================================================================================
+// VFS
+
+// -----------------------------------------------------------------------------
+// Code to wrap SQLite's native VFS so that it can be rooted in some `kj::Directory`, where that
+// directory points at a real disk directory.
+//
+// A native disk `kj::Directory` -- at least on Unix -- wraps an open file descriptor, pointing at
+// a directory. It does NOT keep track of the directory's path on disk. In fact, the directory can
+// be moved or renamed, and `kj::Directory` will continue to point at it.
+//
+// There is no portable way to query the current path of a directory. In order to open files within
+// a directory given only the directory descriptor, you must use syscalls like `openat()`, which
+// take a directory file descriptor to use as the root.
+//
+// SQLite's native VFS, however, is not openat()-aware. Luckily, it _does_ provide the ability to
+// redirect its syscalls to custom implementations. So we can intercept `open()` and make it use
+// `openat()` instead! With a little thread-local hackery, we can make sure to use the desired root
+// directory descriptor from the `kj::Directory`.
+//
+// Of course, SQLite also lets us virtualize the whole filesystem at a higher level. Why go to all
+// the bother to hack it at a low level rather than just implement an entire VFS based on the
+// `kj::Directory` interface? The problem is, SQLite's native VFS contains a ton of code to handle
+// all sorts of corner cases and do things just right. When our files are actually on real disk,
+// we want to leverage all that code. If we can just make it interpret paths differently, then we
```

Comment:
Admittedly I probably went a bit overboard here to make this pretty.

That said, avoiding this complexity in `sqlite.c++` would have required adding complexity in `server.c++`.

First, note that there is no portable way to get a directory's path given an open FD of that directory. On Linux you can sometimes use /proc (though /proc is often not available inside sandboxes). Mac has some other API, but FreeBSD AFAICT does not have such an API at all. Therefore, it would be necessary to change code in server.c++ which passes around `kj::Own<const kj::Directory>` to pass around `kj::Path` instead, which seems disappointing.

Second, `Server` itself takes an abstract `kj::Filesystem&` passed to its constructor. `server-test.c++` passes in an in-memory filesystem while `workerd.c++` uses a real disk filesystem. Nothing in `server.c++` knows which it is using. But if the SQLite interface had to be told explicitly which one it was receiving, then `server.c++` would presumably need to be changed to be aware of it, which in turn erodes trust in the unit test, since it's not testing exactly the same code that runs in production.

Because of all this, I feel that the solution I came up with is the most maintainable approach. It was arguably harder to implement than was warranted, but now that it's done I prefer to keep it.

Reply:
I think I'm okay with this explanation but to summarize my understanding:
We'd probably want to make some sort of top level factory type that is separate from `Server` to be able to use the normal VFS setup. This would be a confusing interface taken together with our existing use of `kj::Filesystem`. It also would probably need to be specialized for sandboxing. Basically, we've accepted higher internal complexity here to better follow expectations for a `kj` ecosystem.

Reply:
I don't think it's specific to KJ.

I would say that using string paths and the default VFS would be a "worse is better" approach -- simpler code, but warty and harder to test, with totally separate code paths needed in server.c++ to support unit testing vs. production. The approach I took instead is a "better is better" approach -- high complexity within `sqlite.c++` itself in order to provide a clean, consistent, object-oriented external-facing interface that supports abstraction and dependency injection, such that server.c++ need not have separate code paths for testing vs. production.